src/hoc/BadgerBase/BadgerBase.js

```diff
@@ -283,6 +308,56 @@ const BadgerBase = (Wrapped: React.AbstractComponent<any>) => {
 			this.setState({ intervalUnconfirmed: intervalUnconfirmedNext });
 		};
 
+		setupWatchInvoice = async () => {
+			const { paymentRequestUrl } = this.props;
+
+			this.ws = new WebSocket(
+				`wss://pay.bitcoin.com/s/${paymentRequestUrl.slice(26)}`
+			);
+
+			this.ws.onmessage = (evt) => {
+				// listen to data sent from the websocket server
+				const initialInvoiceInfo = JSON.parse(evt.data);
+				const invoiceStatus = initialInvoiceInfo.status;
+				if (invoiceStatus === 'paid') {
+					return this.paymentSendSuccess();
+				}
+				if (invoiceStatus === 'expired') {
+					return this.invoiceExpired();
+				}
+
+				// If invoice is not expired or paid, start the timer (add this logic after timer works) TODO
+				// Get current UTC time for timer
+				const now = new Date();
+				const nowUTC = Date.parse(now.toUTCString());
+
+				const invoiceExpiresAt = Date.parse(initialInvoiceInfo.expires);
+
+				let invoiceTimeLeftSeconds = Math.round(
+					(invoiceExpiresAt - nowUTC) / 1000
+				);
+
+				// Prod
+				if (invoiceTimeLeftSeconds > 0) {
+					this.setState({ invoiceTimeLeftSeconds });
+					// start timer
+					const intervalTimerNext = setInterval(() => {
```

Comment:
Potentially starting many times and could lead to a memory leak as this runs on each message.
Instead I would track when the invoice expires, and start a 1 second interval to update the current time in `state` which gets setup in `componentDidMount` and cleaned up in `componentWillUnmount`.  
This way there's a constant 1 second timer, and it only needs to compare to the expiry time

Reply:
c5cd924

On websocket message, time left is stored in `invoiceTimeLeftSeconds` in state. When `invoiceTimeLeftSeconds` in state changes from `null` to `!== null`, timer is started. Interval is cleared on invoice expiration, on payment success, or on `componentWillUnmount()`